Fails on Windows (only) with ESMOKER_FATAL "PackError: Package manager "npm" failed to pack"

For example: https://github.com/DavidAnson/markdownlint-cli2/actions/runs/5993175027

The same configuration passes on macOS and Ubuntu. `--verbose` was not enough to expand the nested `cause` object.

```
Run npm exec --yes -- midnight-smoker --verbose test-invoke-as-cli
💨 midnight-smoker v6.1.1
- Packing current project…
✖ Package manager "npm" failed to pack
FatalError: midnight-smoker failed unexpectedly
    at Smoker.smoke (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\smoker.ts:655:13)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Object.handler (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\cli.ts:423:13) {
  code: 'ESMOKER_FATAL',
  [cause]: {
    error: PackError: Package manager "npm" failed to pack
        at Npm7.pack (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\pm\npm7.ts:137:13)
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
        at Smoker.pack (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\smoker.ts:314:20)
        at Smoker.smoke (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\smoker.ts:618:34)
        at Object.handler (C:\npm\cache\_npx\b66bfc48521ee935\node_modules\midnight-smoker\src\cli.ts:423:13) {
      code: 'ESMOKER_PACK',
      [cause]: [Object]
    }
  }
}
Error: Process completed with exit code 1.
```

--- a/package.json
+++ b/package.json
@@ -124,6 +124,7 @@
     "rewiremock": "3.14.5",
     "sinon": "15.2.0",
     "snap-shot-it": "7.9.10",
+    "strip-ansi": "5.2.0",
     "ts-node": "10.9.1",
     "typescript": "5.2.2",
     "unexpected": "13.2.1",

--- a/src/cli.ts
+++ b/src/cli.ts
@@ -1,7 +1,8 @@
-import {blue, cyan, red, white, yellow} from 'chalk';
+import {blue, cyan, dim, red, white, yellow} from 'chalk';
 import createDebug from 'debug';
 import deepMerge from 'deepmerge';
 import {error, info, warning} from 'log-symbols';
+import {inspect} from 'node:util';
 import ora from 'ora';
 import pluralize from 'pluralize';
 import {hideBin} from 'yargs/helpers';
@@ -267,8 +268,8 @@ async function main(args: string[]): Promise<void> {
                 msg += '…';
                 spinner.succeed(msg);
               })
-              .once(Event.PACK_FAILED, (err) => {
-                spinner.fail(err.message);
+              .once(Event.PACK_FAILED, () => {
+                spinner.fail('Failed while packing!');
                 process.exitCode = 1;
               })
               .once(
@@ -299,8 +300,8 @@ async function main(args: string[]): Promise<void> {
                   spinner.start(msg);
                 },
               )
-              .once(Event.INSTALL_FAILED, (err) => {
-                spinner.fail(err.message);
+              .once(Event.INSTALL_FAILED, () => {
+                spinner.fail('Failed while installing!');
                 process.exitCode = 1;
               })
               .once(Event.INSTALL_OK, ({uniquePkgs}) => {
@@ -418,7 +419,20 @@ async function main(args: string[]): Promise<void> {
                   console.error(`» ${yellow(dir)}`);
                 }
               });
-            await smoker.smoke();
+
+            try {
+              await smoker.smoke();
+            } catch (e) {
+              const err = e as Error;
+              console.error();
+              if (verbose) {
+                console.error(inspect(err, {depth: null, colors: true}));
+              } else {
+                console.error(err.message);
+                console.error(dim('(use --verbose for more details)'));
+              }
+              process.exitCode = 1;
+            }
           }
         },
       )
@@ -435,10 +449,8 @@ async function main(args: string[]): Promise<void> {
       .strict()
       .parseAsync();
   } catch (err) {
-    if (verbose) {
-      throw err;
-    }
-    console.error(error, red(err));
+    // I think anything here would happen before we started smoking
+    console.error(err);
     process.exitCode = 1;
   }
 }

--- a/src/error.ts
+++ b/src/error.ts
@@ -4,8 +4,8 @@
  */
 
 import {bold} from 'chalk';
-import {SmokeResults} from './types';
 import {StaticCheckContext} from '.';
+import {SmokeResults} from './types';
 
 /**
  * Options for {@link SmokerError} with a generic `Cause` type for `cause` prop.
@@ -56,6 +56,7 @@ export class InvalidArgError extends SmokerError<{arg?: string}> {
 }
 
 export class NotImplementedError extends SmokerError {
+  public readonly name = 'NotImplementedError';
   constructor(message: string) {
     super(message, {code: 'ESMOKER_NOTIMPLEMENTED'});
   }
@@ -64,6 +65,8 @@ export class NotImplementedError extends SmokerError {
 export class DirCreationError extends SmokerError<{
   error: NodeJS.ErrnoException;
 }> {
+  public readonly name = 'DirCreationError';
+
   constructor(message: string, error: NodeJS.ErrnoException) {
     super(message, {code: 'ESMOKER_DIRCREATION', cause: {error}});
   }
@@ -73,6 +76,7 @@ export class DirDeletionError extends SmokerError<{
   dir: string;
   error: NodeJS.ErrnoException;
 }> {
+  public readonly name = 'DirDeletionError';
   constructor(message: string, dir: string, error: NodeJS.ErrnoException) {
     super(message, {code: 'ESMOKER_DIRDELETION', cause: {dir, error}});
   }
@@ -81,6 +85,7 @@ export class DirDeletionError extends SmokerError<{
 export class PackageManagerIdError extends SmokerError<{
   pmId: string;
 }> {
+  public readonly name = 'PackageManagerIdError';
   constructor() {
     super(
       'Could not find package manager ID; please report this bug at https://github.com/boneskull/midnight-smoker/issues/new',
@@ -93,6 +98,7 @@ export class PackageManagerError extends SmokerError<{
   error: Error;
   pmId: string;
 }> {
+  public readonly name = 'PackageManagerError';
   constructor(message: string, pmId: string, error: Error) {
     super(message, {code: 'ESMOKER_PACKAGEMANAGER', cause: {pmId, error}});
   }
@@ -248,12 +254,6 @@ export class UnknownDistTagError extends SmokerError<{
 }> {
   constructor(message: string, pkgName: string, tag: string) {
     super(message, {code: 'ESMOKER_UNKNOWNDISTTAG', cause: {pkgName, tag}});
-  }
-}
-
-export class FatalError extends SmokerError<{error: Error}> {
-  constructor(message: string, error: Error) {
-    super(message, {code: 'ESMOKER_FATAL', cause: {error}});
   }
 }
 

--- a/src/pm/npm7.ts
+++ b/src/pm/npm7.ts
@@ -1,10 +1,11 @@
+import {red} from 'chalk';
 import createDebug from 'debug';
 import path from 'node:path';
 import type {SemVer} from 'semver';
 import {InstallError, PackError, PackParseError} from '../error';
 import type {InstallManifest} from '../types';
 import type {CorepackExecutor} from './corepack';
-import type {ExecResult} from './executor';
+import type {ExecError, ExecResult} from './executor';
 import {GenericNpmPackageManager} from './npm';
 import type {
   InstallResult,
@@ -83,7 +84,27 @@ export class Npm7 extends GenericNpmPackageManager implements PackageManager {
       installResult = await this.executor.exec(installArgs, {
         cwd: tarballRootDir,
       });
-    } catch (err) {
+    } catch (e) {
+      const err = e as ExecError;
+      let parsed: {error: {summary: string}} | undefined;
+      try {
+        this.debug('Trying to parse stdout: %s', err.stdout);
+        parsed = JSON.parse(err.stdout);
+      } catch {
+        // ignore
+      }
+      if (parsed?.error) {
+        throw new InstallError(
+          `Package manager "${this.name}" failed to install packages:\n\n${red(
+            parsed.error.summary,
+          )}`,
+          this.name,
+          {
+            error: err as Error,
+            output: err.stderr,
+          },
+        );
+      }
       throw new InstallError(
         `Package manager "${this.name}" failed to install packages`,
         this.name,
@@ -133,8 +154,29 @@ export class Npm7 extends GenericNpmPackageManager implements PackageManager {
 
     try {
       packResult = await this.executor.exec(packArgs);
-    } catch (err) {
+    } catch (e) {
+      const err = e as ExecError;
       this.debug('(pack) Failed: %O', err);
+      // in some cases we can get something more user-friendly via the JSON output
+      let parsed: {error: {summary: string}} | undefined;
+      try {
+        this.debug('Trying to parse stdout: %s', err.stdout);
+        parsed = JSON.parse(err.stdout);
+      } catch {
+        // ignore
+      }
+      if (parsed?.error) {
+        throw new PackError(
+          `Package manager "${this.name}" failed to pack:\n\n${red(
+            parsed.error.summary,
+          )}`,
+          this.name,
+          {
+            error: err as Error,
+            output: err.stderr,
+          },
+        );
+      }
       throw new PackError(
         `Package manager "${this.name}" failed to pack`,
         this.name,
@@ -144,6 +186,8 @@ export class Npm7 extends GenericNpmPackageManager implements PackageManager {
       );
     }
 
+    // I am not sure why exitCode would be non-zero here.  keep an eye on this;
+    // it might never be called
     if (packResult.exitCode) {
       this.debug('(pack) Failed: %O', packResult);
       throw new PackError(

--- a/src/pm/npm9.ts
+++ b/src/pm/npm9.ts
@@ -1,8 +1,10 @@
+import {red} from 'chalk';
 import createDebug from 'debug';
 import type {SemVer} from 'semver';
 import {InstallError} from '../error';
 import type {InstallManifest} from '../types';
 import type {CorepackExecutor} from './corepack';
+import type {ExecError} from './executor';
 import {Npm7} from './npm7';
 import type {
   InstallResult,
@@ -79,19 +81,31 @@ export class Npm9 extends Npm7 implements PackageManager {
       installResult = await this.executor.exec(installArgs, {
         cwd: tarballRootDir,
       });
-    } catch (err) {
+    } catch (e) {
+      const err = e as ExecError;
+      let parsed: {error: {summary: string}} | undefined;
+      try {
+        this.debug('Trying to parse stdout: %s', err.stdout);
+        parsed = JSON.parse(err.stdout);
+      } catch {
+        // ignore
+      }
+      if (parsed?.error) {
+        throw new InstallError(
+          `Package manager "${this.name}" failed to install packages:\n\n${red(
+            parsed.error.summary,
+          )}`,
+          this.name,
+          {
+            error: err as Error,
+            output: err.stderr,
+          },
+        );
+      }
       throw new InstallError(
         `Package manager "${this.name}" failed to install packages`,
         this.name,
         {error: err as Error},
-      );
-    }
-    if (installResult.exitCode) {
-      this.debug('(install) Failed: %O', installResult);
-      throw new InstallError(
-        `Package manager "${this.name}" failed to install packages`,
-        this.name,
-        {exitCode: installResult.exitCode},
       );
     }
 

--- a/src/smoker.ts
+++ b/src/smoker.ts
@@ -8,7 +8,6 @@ import StrictEventEmitter from 'strict-event-emitter-types';
 import {
   DirCreationError,
   DirDeletionError,
-  FatalError,
   InvalidArgError,
   PackageManagerError,
   PackageManagerIdError,
@@ -31,10 +30,10 @@ import {
 import {
   CheckContext,
   CheckSeverities,
+  RuleOptions,
   type CheckOptions,
   type RuleCont,
   type StaticCheckContext,
-  RuleOptions,
 } from './rules';
 import {BuiltinRuleConts} from './rules/builtin';
 import {
@@ -670,8 +669,6 @@ export class Smoker extends createStrictEventEmitterClass() {
       }
 
       return smokeResults;
-    } catch (err) {
-      throw new FatalError('midnight-smoker failed unexpectedly', err as Error);
     } finally {
       await this.cleanup();
     }

--- a/test/e2e/cli.spec.ts
+++ b/test/e2e/cli.spec.ts
@@ -2,10 +2,10 @@ import fs from 'node:fs/promises';
 import path from 'node:path';
 import snapshot from 'snap-shot-it';
 import unexpected from 'unexpected';
+import type {RawRunScriptResult} from '../../src/types';
 import {readPackageJson} from '../../src/util';
 import assertions from '../assertions';
 import {execSmoker, fixupOutput} from './helpers';
-import type {RawRunScriptResult} from '../../src/types';
 
 const expect = unexpected.clone().use(assertions);
 
@@ -147,6 +147,79 @@ describe('midnight-smoker', function () {
       });
     });
 
+    describe('when packing fails', function () {
+      const cwd = path.join(__dirname, 'fixture', 'pack-error');
+      let result: RawRunScriptResult;
+
+      before(async function () {
+        try {
+          await execSmoker([], {
+            cwd,
+          });
+          expect.fail('should have failed');
+        } catch (err) {
+          result = err as RawRunScriptResult;
+        }
+      });
+
+      it('should provide a reason [snapshot]', async function () {
+        snapshot(fixupOutput(result.stderr));
+      });
+
+      describe('when in verbose mode', function () {
+        before(async function () {
+          try {
+            await execSmoker(['--verbose'], {
+              cwd,
+            });
+            expect.fail('should have failed');
+          } catch (err) {
+            result = err as RawRunScriptResult;
+          }
+        });
+
+        it('should provide more detail [snapshot]', async function () {
+          snapshot(fixupOutput(result.stderr));
+        });
+      });
+    });
+
+    describe('when installation fails', function () {
+      const cwd = path.join(__dirname, 'fixture', 'install-error');
+      let result: RawRunScriptResult;
+
+      before(async function () {
+        try {
+          result = await execSmoker([], {
+            cwd,
+          });
+        } catch (err) {
+          result = err as RawRunScriptResult;
+        }
+      });
+
+      it('should provide a reason [snapshot]', async function () {
+        snapshot(fixupOutput(result.stderr));
+      });
+
+      describe('when in verbose mode', function () {
+        before(async function () {
+          try {
+            await execSmoker(['--verbose'], {
+              cwd,
+            });
+            expect.fail('should have failed');
+          } catch (err) {
+            result = err as RawRunScriptResult;
+          }
+        });
+
+        it('should provide more detail [snapshot]', async function () {
+          snapshot(fixupOutput(result.stderr));
+        });
+      });
+    });
+
     describe('option', function () {
       describe('--version', function () {
         it('should print version and exit', async function () {
@@ -226,8 +299,8 @@ describe('midnight-smoker', function () {
           before(async function () {
             try {
               await execSmoker(['smoke', '--json', '--no-checks'], {cwd});
-            } catch (e) {
-              result = e as RawRunScriptResult;
+            } catch (err) {
+              result = err as RawRunScriptResult;
             }
           });
 

--- a/test/e2e/fixture/install-error/package.json
+++ b/test/e2e/fixture/install-error/package.json
@@ -1,0 +1,8 @@
+{
+  "name": "install-error",
+  "version": "1.0.0",
+  "description": "this should be hard to install",
+  "peerDependencies": {
+    "install-error": "2.0.0"
+  }
+}

--- a/test/e2e/fixture/pack-error/package.json
+++ b/test/e2e/fixture/pack-error/package.json
@@ -1,0 +1,4 @@
+{
+  "name": "pack-error",
+  "description": "npm wants a version field or else"
+}

--- a/test/e2e/helpers.ts
+++ b/test/e2e/helpers.ts
@@ -2,10 +2,12 @@
  * E2E test harness helpers
  * @module
  */
+
 import createDebug from 'debug';
 import {node as execa, type NodeOptions} from 'execa';
 import path from 'node:path';
 import {inspect} from 'node:util';
+import stripAnsi from 'strip-ansi';
 import type {RawRunScriptResult} from '../../src';
 
 const debug = createDebug('midnight-smoker:test:e2e');
@@ -42,17 +44,20 @@ export function dump(obj: any): void {
 /**
  * Strips a bunch of stuff out of a CLI output string that is dependent upon
  * local paths and current versions, etc; stuff that isn't suitable for snapshots.
- * @param stdout String of CLI output
+ * @param str String of CLI output; usually either `stdout` or `stderr`
  * @param stripPmVersions If true, replace `version` in `(npm|yarn|pnpm)@<version>` with the string `<version>`.
  * @returns Fixed output
  */
-export function fixupOutput(stdout: string, stripPmVersions = true) {
-  let result = stdout
+export function fixupOutput(str: string, stripPmVersions = true) {
+  let result = stripAnsi(str)
     // strip the paths to npm/node/corepack in command
     .replace(
       /(?:[^" ]+?)(\/(\.)?bin\/(node|npm|corepack)(?:\.exe|\.cmd)?)/g,
       '<path/to/>$1',
     )
+    .replace(/--pack-destination=\S+/g, '--pack-destination=<path/to/dir>')
+    // I hope I don't get a CVE for this one
+    .replace(/(?<=\s)\S+?\.(log|tgz|txt)/g, '<path/to/some>.$1')
     // strip the versions since it will change
     .replace(/midnight-smoker v\d+\.\d+\.\d+/g, 'midnight-smoker v<version>')
     .replace(/--version\\n\\n\d+\.\d+\.\d+/g, '--version\\n\\n<version>')

